dotnet/runtime
# Adding CPUID for AVX10.2

### Overview
This PR tracks addition of CPUID for AVX10.2 to runtime. We are following the [spec doc](https://cdrdv2-public.intel.com/828965/361050-intel-avx10.2-spec.pdf) to add the CPUID for AVX10.2. Following things have been modified as a part of this PR

1. Adding a new ISA ```InstructionSet_Avx10v2``` and ```InstructionSet_Avx10v2_V512```
2. Adding a new environment variable ```EnableAvx10v2``` to JIT.
3. Handle enabling/disabling of AVX10 ISA in a careful manner. ```EnableAvx10v1``` will take priority over ```EnableAvx10v2``` i.e. if ```Avx10v1``` is disabled, ```Avx10v2``` will be disabled too.
4. Remove checks for CPUID for AVX10v1_V128 since bit 17 will imply V128 and V256 support i.e. minimum support on AVX10.
![image](https://github.com/user-attachments/assets/1b3e8474-5368-4dd5-aa6f-e2074a3ce110)


### Testing Correctness of Logic
Since we don't have the DMR hardware available right now, the testing for this PR was done on internal SDE. Following scenarios were tested on internal SDE

![image](https://github.com/user-attachments/assets/0c18c50e-4249-4371-a6e1-f0b08f8b84de)

1. Add debug prints to check the value of ```cpufeatures```  after setting all the ISA flags.
![image](https://github.com/user-attachments/assets/237dc45d-efb2-421b-8aa9-297643a12986)
![image](https://github.com/user-attachments/assets/c9812e6b-57ac-431d-8aec-3b1b4021134d)
![image](https://github.com/user-attachments/assets/52e1efeb-2a2d-4571-a3ba-9b9dd91430b3)

2. Enable ```AVX10v1``` and ```AVX10v2``` by default and do the same. 
![image](https://github.com/user-attachments/assets/382dbc8a-4953-4aa5-aff6-3eac06e96723)

3. Disable ```Avx10v1``` and check if it disables ```Avx10v2```. (it should disable ```AVX10v2```)
![image](https://github.com/user-attachments/assets/880c621c-3318-4308-9335-f5d53d989c23)

4. Disable ```Avx10v2``` and verify that ```Avx10v1``` stays enabled.
![image](https://github.com/user-attachments/assets/b09f0aab-7318-4496-9f9f-3ceff5a523b7)

### Testing JIT test suite
----
The testing was done using the Rex2 unit testing framework. All the changes made for testing CPUID changes are present in [this repo](https://github.com/khushal1996/runtime/tree/kcm-avx102-cpuid-sde-testing)
 
A brief overview --> 
- Unit tests ran to make sure no existing test is breaking
- SDE embedded in test framework
- Test ran with Intel internal SDE using ```dmr``` machine

**Step 1**: Run all tests without SDE
![image](https://github.com/user-attachments/assets/71656dc6-82cc-4147-82e3-9dfd5158b410)


**Step 2**: Running all tests on the **current branch** with ```sde -dmr``` and ```DOTNET_EnableAVX10v2=0```
![image](https://github.com/user-attachments/assets/d030d15e-fe96-4286-bc48-65a35f75af3f)


**Step 3**: Running all tests on the **current branch** with ```sde -dmr```
![image](https://github.com/user-attachments/assets/eae59c26-b209-4a6e-a7bf-83ac7a146467)


### Testing ASMDIFFS
----
Using ```jit-diff``` method, this testing step compares assembly between the ```main``` and the current branch and makes sure that we are not changing any existing scenarios.

Note - we cannot use superpmi here since we are adding a new CPUID and hence the ```jiteeversionguid``` has changed which would lead to wrong results from superpmi.

![image](https://github.com/user-attachments/assets/e0aea610-f8e7-48b7-b048-fb54407427ec)
